### PR TITLE
[#33612] Add a extension installer script feature

### PR DESCRIPTION
**Pull request by wilsonge**
http://joomlacode.org/gf/project/joomla/tracker/?action=TrackerItemEdit&tracker_item_id=33612&start=0
## Feature Basics

This is a helper for developers using install scripts in their extensions it allows you to:
- Get and set parameters (method getParams and setParam available for this - you can also get all instances of a module with getInstances to help)
- Stop users from downgrading versions (set allowDowngrades class var to true)
- Move CLI scripts into the CLI folder (by putting the path to the cli scripts in the class var array)
- Check for a minimum PHP and Joomla version (by setting the strings for the minimum versions in the relevant class var)
- Remove out of date files and folders (by putting the path to the files in the class var array)
## Usage

Simply by extending the helper class and setting the extension name in the form of com_foobar, mod_foobar etc. in the `$this->extension` class var
## Testing

Try using the feature in extensions - as this is a utility class for developers there's no easy visible effects for user testing

### CODE REVIEW

**Comment by compojoom** [comment]
Don't you think that we should find a different name for this class? 
the supposed way we should use this is by extending, but extending from a helper class doesn't seem right to me. A helper class should not be extended from - it is normally something that we have just because we have no clue where to put the functionality. 
I would actually prefer if you move this to the installer folder and call the class something like installerscriptbase??? 
Other than that - great! A lot of us have written their own installers. You are covering a lot of things. What would make it even better is if you had plugin, module & library installers there :)
Something like here: https://github.com/compojoom/lib_compojoom/blob/master/source/libraries/compojoom/installer/installer.php#L110

**Comment by wilsonge** [comment]
I'm easy as to the name of the class and it's location. Yes you're right it might well be better in ```libraries/cms/installer```.

I considered the library, module and plugin installers but in my opinion core should be persuading people to use packages - they're getting much better in terms of core support since the standardisation we did in 3.4. What do you think?

**Comment by mbabker** [comment]
@wilsonge see https://github.com/wilsonge/joomla-cms/pull/13

The helper class is renamed to `JInstallerScript` and I've updated 3.4 references to 3.5.

**Comment by mbabker** [comment]
> What would make it even better is if you had plugin, module & library installers there

I tend to agree with @wilsonge on using packages instead of extending a specific adapter to add more extensions.  Admittedly there are some weaknesses in the package adapter that need to be worked on to make it more enticing for developers (i.e. one failed extension install aborts the whole thing leaving a partial package install), but I think our focus for core should be moving in that direction.

File: libraries/cms/installer/script.php [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+
+defined('_JEXEC') or die;
+
+jimport('joomla.filesystem.file');
+jimport('joomla.filesystem.folder');
+
+/**
+ * Base install script for use by extensions providing helper methods for common behaviours.
+ *
+ * @since  3.5
```
Comment (810) [inline libraries/cms/installer/script.php]: Please update all since to 3.6

File: libraries/cms/installer/script.php [hunk 2]
```diff
@@ -132,7 +132,7 @@ public function preflight($type, $parent)
 		}
 
 		// Abort if the extension being installed is not newer than the currently installed version
-		if ($type == 'Update' && $this->allowDowngrades)
+		if (strtolower($type) == 'Update' && !$this->allowDowngrades)
```
Comment (wojsmol) [inline libraries/cms/installer/script.php]: IMHO this should be
`if (strtolower($type) == 'update' && !$this->allowDowngrades)`

Comment (wilsonge) [inline libraries/cms/installer/script.php]: Fixed

File: libraries/cms/installer/script.php (hunk 1 above)
Comment (wilsonge) [inline libraries/cms/installer/script.php]: Done

**Comment by Kubik-Rubik** [comment]
Thank you @wilsonge, great addition!

**Pull request merged**